### PR TITLE
feat(rxCheckbox): add rxCheckbox component

### PR DESCRIPTION
**Pull request by CITguy**
Per necessity of #884, this directive can stand alone for use elsewhere in the framework.

### CODE REVIEW

**Comment by parlarjb** [comment]
Could you include some screenshots?

**Comment by CITguy** [comment]
screen shot 2015-05-06 at 9 56 33 am

**Comment by CITguy** [comment]
Rebased

**Comment by megalord** [comment]
I'm curious, why are the attributes being documented in the demo markup as opposed to via ngdocs?  That seems to be the standard from what I've seen.

File: src/rxCheckbox/rxCheckbox.less [hunk 1]
```diff
@@ -0,0 +1,99 @@
+/*
+ * rxCheckbox
+ */
+.rxCheckbox {
+  display: inline-block;
+  width: @rxCheckbox-size; // OK
```
Comment (megalord) [inline src/rxCheckbox/rxCheckbox.less]: Can you remove all of the `// OK` comments?

Comment (CITguy) [inline src/rxCheckbox/rxCheckbox.less]: Ah, good call.

File: src/rxCheckbox/rxCheckbox.page.js [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                }
+            });
+        }
+    },
+
+    /**
+     * @function
+     * @description make sure checkbox is deselected/unchecked
+     */
+    deselect: {
```
Comment (megalord) [inline src/rxCheckbox/rxCheckbox.page.js]: The `rxForm` version of a checkbox uses "unselect" instead of "deselect".  While I'm not partial to either, it may be confusing for anyone migrating to have to change their tests.  It would be easier if all they had to do is change
```javascript
rxForm.checkbox.initialize(...)
```
to
```javascript
rxCheckbox.initialize(...)
```

**Comment by CITguy** [comment]
@megalord, I need to apply a styling to the first cell in the row so that the value doesn't wrap.  I cannot do this via markdown.  I addition, I wanted to keep non-demo markup to bhe bottom so that when a consumer looks at the "Markup" tab, they see the demo code first.

File: src/rxCheckbox/rxCheckbox.page.js (hunk 2 above)
Comment (CITguy) [inline src/rxCheckbox/rxCheckbox.page.js]: Keep in mind, that this PR is to support upcoming changes to rxForm.

Comment (CITguy) [inline src/rxCheckbox/rxCheckbox.page.js]: Also, http://english.stackexchange.com/questions/18465/unselect-or-deselect

File: src/rxCheckbox/rxCheckbox.spec.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+        inject(function ($location, $rootScope, $compile) {
+            rootScope = $rootScope;
+            scope = $rootScope.$new();
+            compile = $compile;
+        });
+
+        el = helpers.createDirective(validTemplate, compile, scope);
+    });
+
+    /* reserved for potential future specs */
```
Comment (megalord) [inline src/rxCheckbox/rxCheckbox.spec.js]: I'm not a big fan of having a spec file without any tests in it.  I say either add tests for the watch on `ng-disabled` or remove the file entirely.

Comment (CITguy) [inline src/rxCheckbox/rxCheckbox.spec.js]: I only kept it in case someone pointed out something that required testing in this review. If there's nothing worth testing, then I can remove it. I'll defer to @Droogans on this one.

File: src/rxCheckbox/rxCheckbox.page.js (hunk 2 above)
Comment (CITguy) [inline src/rxCheckbox/rxCheckbox.page.js]: Should rxForm use rxCheckbox, then `rxForm.checkbox.initialize(...)` will initialize an rxCheckbox item.

**Comment by megalord** [comment]
@CITguy ngdocs doesn't use markdown - it's generated from the comments that are placed before each `.directive`, `.service`, etc.

File: src/rxCheckbox/rxCheckbox.page.js (hunk 2 above)
Comment (CITguy) [inline src/rxCheckbox/rxCheckbox.page.js]: Sorry, I misunderstood.  With the upcoming changes to rxForm, there's going to be a considerable update. `rxForm.checkbox.initialize(...)` only applies to old rxForm functionality (using native checkboxes).  As developers transition to using the new rxForm functionality (using newly styled checkboxes), they would naturally have to transition to the new page objects as well.

Comment (megalord) [inline src/rxCheckbox/rxCheckbox.page.js]: Yeah, there will be some big changes to rxForm, and my concern is that something like this is not obvious and will be lost amidst the other changes, so the tests will break in unexpected places.  Perhaps as the larger change lands, it would be wise to have a migration guide where the big picture changes as well as details like this can be documented.

Comment (CITguy) [inline src/rxCheckbox/rxCheckbox.page.js]: I think we can revisit this topic in the rxForm updates, if needed.

**Comment by CITguy** [comment]
Added "DO NOT MERGE" because functionality is needed to remove the wrapper markup on a destroy (thanks for pointing this out @megalord). This is needed when an `ng-if` or similar removes the `<input rx-checkbox ... />` element from the DOM.

File: src/rxCheckbox/rxCheckbox.page.js (hunk 2 above)
Comment (Droogans) [inline src/rxCheckbox/rxCheckbox.page.js]: Since I'm the one responsible for this odd naming convention, I feel like I should at least explain why I went with unselect vs. deselect. At first, there was `select` for dropdowns and radio inputs, and `check` for checkboxes.

I didn't like the variation in language. It made it harder for me to remember what to call. So I switched everything over to `select`, and in the process, saw `uncheck` and thought...."should this be `unselect`? What about `deselect`?". After thinking about it, I went with something that worked everywhere (since there was no `decheck`), ending up settling on `unselect`, which may need to change.

TL;DR - I'm all for switching everything over to `deselect`, since there's no more variation between verbs.

Comment (Droogans) [inline src/rxCheckbox/rxCheckbox.page.js]: Also, @megalord, @CITguy and I discussed what'll likely happen to the page objects once all the rxForm components get broken out of it. I think we'll be using `rxForm` as a "require everything please" page object, so it'll keep backwards compatibility. It'll also make pulling in every form input type less lines of code (one `require` instead of three or four).

**Comment by CITguy** [comment]
* Rebased
* removed `rxCheckbox.spec.js` (no functionality to test)
* added `$destroy` support
* updated midways and page object
* updated demo and documentation

**Comment by CITguy** [comment]
rxcheckbox-local-20150508

**Comment by CITguy** [comment]
*unchecked (invalid)*
screen shot 2015-05-08 at 11 25 46 am

*checked (valid), shows next checkbox via `ng-show` logic*
screen shot 2015-05-08 at 11 25 54 am

*both checked and valid*
screen shot 2015-05-08 at 11 26 03 am

**Comment by CITguy** [comment]
**`$destroy` support**
*unchecked, second checkbox is present*
screen shot 2015-05-08 at 11 26 11 am

*checked, second checkbox (and stylistic markup) removed from DOM*
screen shot 2015-05-08 at 11 26 17 am

File: src/rxCheckbox/README.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+## Show/Hide
+
+If you wish to show/hide your `rxCheckbox` element (and its label), we recommend placing the element (and its label) inside of a
+`<div>` or `<span>` wrapper, and performing the show/hide logic on the wrapper.
+
+```html
+<span ng-show="isShown">
+  <input rx-checkbox
+         type="checkbox"
```
Comment (parlarjb) [inline src/rxCheckbox/README.md]: Could the `rx-checkbox` directive automatically set the `type="checkbox"`, in the `compile` step?

File: src/rxCheckbox/README.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+         id="chkDemo"
+         ng-model="chkDemo" />
+  <label for="chkDemo">Label for Demo Checkbox</label>
+</span>
+```
+
+It is highly recommended that you use `ng-show` and `ng-hide` for purposes of display logic. Because of the way that `ng-if` and `ng-switch`
+directives behave with scope, they may introduce unnecessary complexity in your code.
```
Comment (parlarjb) [inline src/rxCheckbox/README.md]: :+1:

File: src/rxCheckbox/rxCheckbox.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
+angular.module('encore.ui.rxCheckbox', [])
+.directive('rxCheckbox', function () {
+    return {
+        restrict: 'A',
+        scope: {
+            ngDisabled: '=?'
```
Comment (parlarjb) [inline src/rxCheckbox/rxCheckbox.js]: This doesn't cause problems with the existing `ngDisabled` directive that ships with Angular?

File: src/rxCheckbox/README.md (hunk 4 above)
Comment (CITguy) [inline src/rxCheckbox/README.md]: Unfortunately, I cannot use both `compile` and `link` in the DDO. Considering that I need to perform DOM manipulation, I need to keep `link`. I tried to automatically set the `type="checkbox"` in the link function, but for some reason, it breaks binding with the scope.

File: src/rxCheckbox/rxCheckbox.js (hunk 6 above)
Comment (CITguy) [inline src/rxCheckbox/rxCheckbox.js]: Nope, just watches the value so that a `scope.$watch()` can add/remove a `disabled` attribute on the wrapper element.

File: src/rxCheckbox/README.md (hunk 4 above)
Comment (CITguy) [inline src/rxCheckbox/README.md]: In addition, if for some reason that the directive doesn't bind properly to the DOM, the `type="checkbox"` will act as a fallback for graceful degradation.

File: src/rxCheckbox/rxCheckbox.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        },
+        link: function (scope, element) {
+            var wrapper = '<div class="rxCheckbox"></div>';
+            var fakeCheckbox = '<div class="fake-checkbox">' +
+                    '<div class="tick fa fa-check"></div>' +
+                '</div>';
+
+            element.wrap(wrapper);
+            element.after(fakeCheckbox);
+
```
Comment (parlarjb) [inline src/rxCheckbox/rxCheckbox.js]: Rather that doing `element.parent()` on every $watch call below, could you do `wrapper = angular.element(wrapper)`,
and then inside the `$watch` you an do `wrapper.attr('disabled', true);` and `wrapper.removeAttr('disabled')`.
Then in the `$destroy` you can also do `wrapper[0].remove();`

File: src/rxCheckbox/README.md (hunk 4 above)
Comment (megalord) [inline src/rxCheckbox/README.md]: Just curious, are you opposed to returning the `link` function from `compile`?

Comment (CITguy) [inline src/rxCheckbox/README.md]: I had a chat with @parlarjb about it. He convinced me to convert to using `compile` and returning the function. I'll be submitting an update to the PR soon.

File: src/rxCheckbox/rxCheckbox.js (hunk 7 above)
Comment (CITguy) [inline src/rxCheckbox/rxCheckbox.js]: ah, I will do that. much cleaner.

**Comment by CITguy** [comment]
Updated per comments:

* using `compile` rather than `link` so that the directive automatically sets `type="checkbox"`
* applying logic to wrapper element rather than `element.parent()` repetition

Additionally, I reorganized some of the documentation to make it easier to consume and navigate. The examples still work same as seen in previous screenshots.

rxcheckbox-local-20150511-1

**Comment by parlarjb** [comment]
Could you put back in `.spec.js` to ensure that `type="checkbox"` is being correctly added, as well as checking that your custom `ng-disabled` logic is being applied?

**Comment by CITguy** [comment]
@parlarjb Midways have been updated to test `ng-disabled` and `type="checkbox"` functionality. I didn't feel like this warranted the need for a spec file as the directive isn't adding any functionality on `scope` or the controller. It merely manipulates the DOM based on data binding. @Droogans do you have any input on testing via unit tests versus midways?

**Comment by parlarjb** [comment]
LGTM

**Comment by CITguy** [comment]
Based on feedback for `rxRadio`, I need to add logic to support the native `disabled` attribute in addition to `ng-disabled`.

**Comment by CITguy** [comment]
* added support for native `disabled` attribute

*Checkbox States - updated to include native `disabled` examples*
screen shot 2015-05-13 at 11 29 57 am

File: src/rxCheckbox/rxCheckbox.js [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+                element.wrap(wrapper);
+                element.after(fakeCheckbox);
+
+                // apply/remove disabled attribute so we can
+                // apply a CSS selector to style sibling elements
+                if (element.attr('disabled')) {
+                    wrapper.addClass(disabledClass);
+                }
+                if (scope.ngDisabled) {
```
Comment (megalord) [inline src/rxCheckbox/rxCheckbox.js]: Should this be `attr.ngDisabled`? If the expression passed to `ng-disabled` is initially false, the watch should still be applied.

Comment (CITguy) [inline src/rxCheckbox/rxCheckbox.js]: I see what you're saying. I was trying to check for the presence of the variable, not its value. I'll get a quick update submitted.

File: src/rxCheckbox/rxCheckbox.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+                element.wrap(wrapper);
+                element.after(fakeCheckbox);
+
+                // apply/remove disabled attribute so we can
+                // apply a CSS selector to style sibling elements
+                if (attrs.disabled) {
+                    wrapper.addClass(disabledClass);
+                }
+                if (angular.isDefined(scope.ngDisabled)) {
```
Comment (CITguy) [inline src/rxCheckbox/rxCheckbox.js]: @megalord, how's this look?

Comment (megalord) [inline src/rxCheckbox/rxCheckbox.js]: Well, this doesn't work for a value of `undefined`, which would work on a normal checkbox input.  At this point though, I feel I'm getting too nit-picky, so feel free to keep it as is unless someone else chimes in.

Comment (parlarjb) [inline src/rxCheckbox/rxCheckbox.js]: `if (_.has(attrs, 'ngDisabled')) {` would probably be the most correct way to do it

**Comment by CITguy** [comment]
Updated to use `if (_.has(attrs, 'ngDisabled'))`

**Comment by megalord** [comment]
LGTM
